Neo4j.Driver/Neo4j.Driver/Internal/Helpers/TemporalHelpers.cs

```diff
@@ -16,92 +16,260 @@
+        private const long NanosPerHour = NanosPerMinute * MinutesPerHour;
+        private const long Days0000To1970 = (DaysPerCycle * 5L) - (30L * 365L + 7L);
+        private const int DaysPerCycle = 146_097;
+        private const int NanosecondsPerTick = 100;
 
-        public static long DaysSinceEpoch(this DateTime date)
+        public static long ToNanoOfDay(this IHasTimeComponents time)
         {
-            return (long)date.Subtract(Epoch).TotalDays;
+            return (time.Hour * NanosPerHour) + (time.Minute * NanosPerMinute) + (time.Second * NanosPerSecond) + time.Nanosecond;
         }
 
-        public static long NanosOf(this TimeSpan time)
+        public static long ToEpochSeconds(this IHasDateTimeComponents dateTime)
         {
-            return time.Ticks * NanosecondsPerTick;
+            var epochDays = dateTime.ToEpochDays();
+            var timeSeconds = dateTime.ToSecondsOfDay();
+
+            return epochDays * SecondsPerDay + timeSeconds;
         }
 
-        public static long SecondsSinceEpoch(long ticks)
+        public static int ToSecondsOfDay(this IHasTimeComponents time)
         {
-            return (ticks / TimeSpan.TicksPerSecond) - EpochSeconds;
+            return (time.Hour * SecondsPerHour) + (time.Minute * SecondsPerMinute) + time.Second;
         }
 
-        public static int NanosOfSecond(long ticks)
+        public static long ToEpochDays(this IHasDateComponents date)
         {
-            return (int)((ticks % TimeSpan.TicksPerSecond) * NanosecondsPerTick);
+            return ComputeEpochDays(date.Year, date.Month, date.Day);
+        }
+
+        public static CypherTime NanoOfDayToTime(long nanoOfDay)
+        {
+            ComponentsOfNanoOfDay(nanoOfDay, out var hour, out var minute, out var second, out var nanosecond);
+
+            return new CypherTime(hour, minute, second, nanosecond);
+        }
+
+        public static CypherDate EpochDaysToDate(long epochDays)
+        {
+            ComponentsOfEpochDays(epochDays, out var year, out var month, out var day);
+
+            return new CypherDate(year, month, day);
+        }
+
+        public static CypherDateTime EpochSecondsAndNanoToDateTime(long epochSeconds, int nano)
+        {
+            var epochDay = FloorDiv(epochSeconds, SecondsPerDay);
+            var secondsOfDay = FloorMod(epochSeconds, SecondsPerDay);
+            var nanoOfDay = secondsOfDay * NanosPerSecond + nano;
+
+            ComponentsOfEpochDays(epochDay, out var year, out var month, out var day);
+            ComponentsOfNanoOfDay(nanoOfDay, out var hour, out var minute, out var second, out var nanosecond);
+
+            return new CypherDateTime(year, month, day, hour, minute, second, nanosecond);
         }
 
-        public static long NanosOf(int hour, int minute, int second, int nanoOfSecond)
+        private static long ComputeEpochDays(int year, int month, int day)
```

Comment:
@lutovich Can you also have a look at this method in this class from https://github.com/ThreeTen/threetenbp?